test/benchmark/cuda_driver.cpp

```diff
@@ -0,0 +1,197 @@
+        return CU_TARGET_COMPUTE_80;
+    case 86:
+        return CU_TARGET_COMPUTE_86;
+    default:
+        throw std::runtime_error("Unsupported compute architecture");
+    }
+}
+
+void CUDADriver::init(const codegen::Platform& platform, BenchmarkInfo* benchmark_info) {
+    // CUDA initialization
+    checkCudaErrors(cuInit(0));
+    checkCudaErrors(cuDeviceGetCount(&device_info.count));
+    checkCudaErrors(cuDeviceGet(&device, 0));
+
+    char name[128];
+    checkCudaErrors(cuDeviceGetName(name, 128, device));
+    device_info.name = name;
+    logger->info("Using CUDA Device [0]: {}"_format(device_info.name));
+
+    checkCudaErrors(cuDeviceGetAttribute(&device_info.compute_version_major,
+                                         CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR,
+                                         device));
+    checkCudaErrors(cuDeviceGetAttribute(&device_info.compute_version_minor,
+                                         CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR,
+                                         device));
+    logger->info("Device Compute Capability: {}.{}"_format(device_info.compute_version_major,
+                                                           device_info.compute_version_minor));
+    if (device_info.compute_version_major < 2) {
+        throw std::runtime_error("ERROR: Device 0 is not SM 2.0 or greater");
+    }
+
+    // Load the external libraries modules to the NVVM program
+    // Currently only libdevice is supported
+    link_libraries(*module, benchmark_info);
+
+    // Compile the program
+    logger->info("Compiling the LLVM IR to PTX");
+
+    // Optimize code for nvptx including the wrapper functions and generate PTX
+    const auto opt_level_codegen = benchmark_info ? benchmark_info->opt_level_codegen : 0;
+    utils::optimise_module_for_nvptx(platform, *module, opt_level_codegen, ptx_compiled_module);
+    utils::save_ir_to_ll_file(*module,
+                              benchmark_info->output_dir + "/" + benchmark_info->filename +
+                                  "_benchmark");
+    if (benchmark_info) {
+        print_string_to_file(ptx_compiled_module,
+                             benchmark_info->output_dir + "/" + benchmark_info->filename + ".ptx");
+    }
+
+    // Create driver context
+    checkCudaErrors(cuCtxCreate(&context, 0, device));
+
+    // Create module for object
+    logger->info("Loading PTX to CUDA module");
+    const unsigned int jitNumOptions = 5;
+    CUjit_option* jitOptions = new CUjit_option[jitNumOptions];
+    void** jitOptVals = new void*[jitNumOptions];
+
+    // set up size of compilation log buffer
+    jitOptions[0] = CU_JIT_INFO_LOG_BUFFER_SIZE_BYTES;
+    size_t jitLogBufferSize = 1024 * 1024;
+    jitOptVals[0] = (void*) jitLogBufferSize;
+
+    // set up pointer to the compilation log buffer
+    jitOptions[1] = CU_JIT_INFO_LOG_BUFFER;
+    char* jitLogBuffer = new char[jitLogBufferSize];
+    jitOptVals[1] = jitLogBuffer;
+
```

Comment:
Objects explicitly allocated like `jitLogBuffer` and `jitOptions` are internally freed by llvm?